### DISCUSSION TITLE
Truffle test ERC20 token transfers

### DISCUSSION THREAD

**Discussion by PijiDev** [ConsenSys-archive/truffle#6011] [General Q&A]
I'm writing my smart contract tests in JS and using Truffle. I want to test a specific function in my contract that requires a token address and contract address as parameters:

```
address _asset, //address base token, like USDC
        address _payee,
        uint8 _dueDate,
        uint256 _amount,
        bytes calldata _paymentReference,
        address _cometAddress
        )
```

When I just pass the mainnet USDC token address and mainnet Comet Address, the test errors out:

```
it("should be able to make an ERC20 payment", async () => {
    await instance.payInvoiceERC20(
      "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48", //USDC mainnet token address
      accounts[8], //setting a payee
      2,
      10000000, //10 USDC, 6 decimals for USDC token
      "0x494e56332d32343034", //payment ref. in bytes
      "0xc3d688B66703497DAA19211EEdff47f25384cdc3" //Comet Address USDC mainnet
    )
  })
```

Error message: "exited with an error (status 0). Reason given: Address: call to non-contract."

From what I read online and SO answers from 2018-2019, I understand I need to somehow instantiate the ERC20 token contract (+ Comet contract). What is the best and fastest way?
Could anyone give me a real example?

**Comment by pclaesen** [comment]
Important note:
The Comet address will only accept the correct USDC token address. I can't just use a mock ERC20 contract.
I can fork Goerli and run the `truffle test` against that fork, so the USDC and Comet contracts are already available.

**Reply by FlaviuWeb3** [reply]
Yes, you're correct that you need to instantiate the ERC20 token contract and Comet contract in order to test your payInvoiceERC20 function.

To do this, you can use Truffle's built-in contract abstraction layer to create instances of the contracts. Here's an example of how you can modify your test code to instantiate the USDC token contract and Comet contract:

```
const USDC = artifacts.require("USDC"); // replace with your actual USDC contract name
const Comet = artifacts.require("Comet"); // replace with your actual Comet contract name

it("should be able to make an ERC20 payment", async () => {
  const usdc = await USDC.at("0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"); // instantiate USDC contract
  const comet = await Comet.at("0xc3d688B66703497DAA19211EEdff47f25384cdc3"); // instantiate Comet contract

  await instance.payInvoiceERC20(
    usdc.address,
    accounts[8],
    2,
    10000000,
    "0x494e56332d32343034",
    comet.address
  );
});
```
In this example, we first import the contract artifacts for the USDC token and Comet contracts using artifacts.require. Then, we instantiate these contracts using await USDC.at and await Comet.at, passing in the mainnet addresses for these contracts.

Finally, we call instance.payInvoiceERC20 with the instantiated contract addresses as arguments.

Note that you may need to modify the code above to match the actual contract names and addresses that you're using in your project.

I hope this helps!

**Reply by pclaesen** [reply]
Hey @FlaviuWeb3, thanks a lot for this! Make more sense now.
Will have to check how to get the correct contract code for the USDC and Comet addresses.
